### PR TITLE
Feat  add pivot transform to mo.UI.dataframe #7345

### LINKED ISSUE #7345: New dataframe transform: Pivot
### Description

I would like to be able to create pivot tables.

### Suggested solution

Using the transform, the user would be able to configure the index, columns, values, and agg functions.

### Are you willing to submit a PR?

- [x] Yes

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by heizerbalazs**
## 📝 Summary

Fixes #7345

### Backend

- Created handler and types for `PivotTransform` in `marimo/_plugins/ui/_impl/dataframes/transforms`
- Ibis backend doesn't support pivot. The handler implementation is using group_by.

### Frontend

- Added icon to mo.ui.dataframe transform panel.
- Added `PivotTransformSchema` with `column_ids, index_column_ids, value_column_ids, aggregation` parameters in `frontend/src/plugins/impl/data-frames/schema.ts`.
- ~~Added effect to collect unique column values so new column names can be generated on the frontend. `frontend/src/plugins/impl/data-frames/panel.ts`~~

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://marimo.io/discord?ref=pr), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

### CODE REVIEW

**Comment by heizerbalazs** [comment]
I have read the CLA Document and I hereby sign the CLA

File: marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py [hunk 1]
```diff
@@ -217,6 +217,32 @@ def generate_where_clause(df_name: str, where: Condition) -> str:
...
+        )
+        pivot_code = f"{df_name}.pivot_table({args}).sort_index(axis=0)"
+        flatten_columns_code = (
+            f"{df_name}.columns = ["
+            f"f\"{{'_'.join(map(str, col)).strip()}}_{agg_func}\" "
+            f'if isinstance(col, tuple) else f"{{col}}_{agg_func}" '
+            f"for col in {df_name}.columns]"
+        )
+        reset_index_code = f"{df_name} = {df_name}.reset_index()"
+        return f"{pivot_code}\n{flatten_columns_code}\n{reset_index_code}"
```
[the remark this answers is not in the document]
Comment (heizerbalazs) [reply marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py:264]: The `python_print_pandas` function returns code with valid syntax. e.g.:
```
df_pandas_next = df_pandas
df_pandas_next = df_pandas_next.pivot_table(index=["Year"], columns=["Origin"], values=["Acceleration"], aggfunc="sum", sort=False, fill_value=0).sort_index(axis=0)
df_pandas_next.columns = [f"{'_'.join(map(str, col)).strip()}_sum" if isinstance(col, tuple) else f"{col}_sum" for col in df_pandas_next.columns]
df_pandas_next = df_pandas_next.reset_index()
df_pandas_next
```

File: marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py [hunk 2]
```diff
@@ -378,6 +404,33 @@ def generate_where_clause_polars(where: Condition) -> str:
...
+                if len(transform.value_column_ids) == 1
+                else f"lambda col: f'{{col.translate(replacements)}}_{transform.aggregation}'"
+            )
+            + f" if col not in {_list_of_strings(transform.index_column_ids)} else col"
+        )
+        rename_code = (
+            'replacements = str.maketrans({"{": "", "}": "", \'"\': "", ",": "_"})\n'
+            f"{df_name} = {df_name}.rename({lambda_code})"
+        )
+        return f"{pivot_code}\n{rename_code}"
```
[the remark this answers is not in the document]
Comment (heizerbalazs) [reply marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py:475]: The `python_print_polars` function returns code with valid syntax. e.g.:
```
df_polars_next = df_polars
df_polars_next = df_polars_next.pivot(on=["Origin"], index=["Year"], values=["Acceleration"], aggregate_function="mean").sort(["Year"])
replacements = str.maketrans({"{": "", "}": "", '"': "", ",": "_"})
df_polars_next = df_polars_next.rename(lambda col: f'Acceleration_{col.translate(replacements)}_mean' if col not in ["Year"] else col)
df_polars_next
```

File: frontend/src/plugins/impl/data-frames/schema.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ export const column_id_array = z
   .array(column_id.describe(FieldOptions.of({ special: "column_id" })))
   .min(1, "At least one column is required")
   .default([])
-  .describe(FieldOptions.of({ label: "Group by columns", minLength: 1 }));
+  .describe(FieldOptions.of({ label: "Columns", minLength: 1 }));
```
Comment (Light2Dark) [inline frontend/src/plugins/impl/data-frames/schema.ts:32]: nice fix!

File: tests/_plugins/ui/_impl/dataframes/test_handlers.py [hunk 4]
```diff
@@ -1184,6 +1185,131 @@ def test_unique(
         )
         assert type(result) is type(df)
 
+    @staticmethod
+    @pytest.mark.parametrize(
```
Comment (Light2Dark) [inline tests/_plugins/ui/_impl/dataframes/test_handlers.py:1189]: i feel this may warrant more tests (can be a smaller dataset), but with different inputs, in particular if the index cols is empty / value cols is empty etc.

File: marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py [hunk 5]
```diff
@@ -378,6 +404,33 @@ def generate_where_clause_polars(where: Condition) -> str:
         column_ids = transform.column_ids
         return f"{df_name}.unique(subset={_list_of_strings(column_ids)}, keep={_as_literal(transform.keep)})"  # noqa: E501
 
+    elif transform.type == TransformType.PIVOT:
```
Comment (Light2Dark) [inline marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py:427]: can we add some tests for `print_code` for pivot?

File: frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts [hunk 6]
```diff
@@ -69,6 +79,36 @@ function handleTransform(
...
+
+        const type =
+          transform.aggregation === "count"
+            ? "int64"
+            : next.get(rawColumn[0] as ColumnId);
+        updated.set(newColumn as ColumnId, type as string);
+      }
+
+      return updated;
+    }
```
Comment (Light2Dark) [inline frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts]: Is it possible for the `handle_pivot` function to do the column names transformation, therefore we don't need to do any custom name handling in the frontend? If so, we can also remove the columnValueCache and all of that complexity here. It'd be better if the transformation logic is in 1 place

**Review by Light2Dark — COMMENTED** [review]
Thanks for the feature!
Is it possible to make the form fields similar to excel. Not the UI, just the naming convention.

1. columns
2. rows
3. values
4. aggregation

Image

File: frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts (hunk 6 above)
Comment (heizerbalazs) [reply frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts]: Thanks for the feedback — I agree with your observation. Other transforms, like group by and aggregation, are already following the same pattern of handling column-name updates on the frontend. I’m open to centralizing the transformation logic, but if it requires a significant refactor, I’d suggest addressing it in a separate issue.

Comment (Light2Dark) [reply frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts]: agreed, I have a PR https://github.com/marimo-team/marimo/pull/7611 that should help (removes all of the frontend columns code). I need to test it more extensively. If that doesn't work, I think your solution will do. Thanks for working on this!

Comment (heizerbalazs) [reply frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts]: Awesome, thank you for looking into this!

Comment (Light2Dark) [reply frontend/src/plugins/impl/data-frames/utils/getUpdatedColumnTypes.ts]: sorry for the merge conflict, let us know if you face trouble integrating

File: marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py (hunk 5 above)
Comment (heizerbalazs) [reply marimo/_plugins/ui/_impl/dataframes/transforms/print_code.py:427]: added test cases for polars and pandas

File: tests/_plugins/ui/_impl/dataframes/test_handlers.py (hunk 4 above)
Comment (heizerbalazs) [reply tests/_plugins/ui/_impl/dataframes/test_handlers.py:1189]: added test cases to cover empty index and value cols. I also made mandatory to select at least one column for values on the ui, otherwise it would blow up the number of columns if the pivot is applied for a wide table.

**Review by Light2Dark — APPROVED** [review]
tq!

**Review by Light2Dark — APPROVED** [review]
